iroh-gateway/src/handlers.rs

```diff
@@ -225,6 +238,54 @@ pub async fn get_handler<T: ContentLoader + std::marker::Unpin>(
     }
 }
 
+#[tracing::instrument(skip(state))]
+pub async fn post_handler<T: ContentLoader + std::marker::Unpin>(
+    Extension(state): Extension<Arc<State<T>>>,
+    // Path(params): Path<HashMap<String, String>>,
+    // Query(query_params): Query<GetParams>,
+    // method: http::Method,
+    http_req: HttpRequest<Body>,
+    // request_headers: HeaderMap,
+) -> Result<GatewayResponse, GatewayError> {
+    // If this gateway is not writable, return a 400 error.
+    if !state.config.writeable_gateway() {
+        return Err(error(
+            StatusCode::BAD_REQUEST,
+            "Not a writable gateway",
+            &state,
+        ));
+    }
+
+    // TODO: check path & headers
+
+    // Helper to convert a anyhow::Error into a http error response.
+    let into_gateway =
+        |err: anyhow::Error| error(StatusCode::INTERNAL_SERVER_ERROR, &err.to_string(), &state);
+
+    // Convert the http body into an AsyncRead
+    let futures_async_read = TryStreamExt::map_err(http_req.into_body(), |_err| {
+        io::Error::new(io::ErrorKind::Other, "Error!")
+    })
+    .into_async_read();
+    let reader = tokio_util::compat::FuturesAsyncReadCompatExt::compat(futures_async_read);
+
+    let cid = state
+        .client
+        .resolver
+        .loader
+        .store_file(reader)
+        .await
+        .map_err(into_gateway)?;
+    let location = format!("ipfs://{}", cid);
+
```

Comment:
after you get the root CID here, you need to call `state.client.try_p2p()?.start_providing().await?`
to actually provide the content to the dht